### ISSUE TITLE
Support drawing Rectangle in any direction

### ISSUE THREAD

**Issue by theashyster** [Leaflet/Leaflet.Editable#87]
Hello,

first of all, great plugin which create a lot of awesome possibilities for us, developers.

I wanted to ask if there are any thoughts given about supporting drawing `Rectangle` in any direction instead of just from the top left to bottom right? I have also played a bit with `Leaflet.Draw` and they support drawing `Rectangle` any way you want and both of these plugins are officially supported by `Leaflet`, so I thought that maybe you guys can put your heads together and support this feature also in `Leaflet.Editable`. :smiley: Or maybe guide me a bit on what should be done to achieve this (I have tried to do it once, by overriding `Rectangle` drawing code and doing my own changes, but something did not work as expected and I put it on hold, but I still need to achieve this at some point).

I also saw that there is this discussion going on https://github.com/Leaflet/Leaflet/issues/4760 about using `Leaflet.Editable` as the core of `Leaflet.Draw`. How is that going to affect the two project moving forward? Is one of them going to be more primary worked on?

Being honest it was a bit confusing to choose which plugin to use when I had to decide that in our project between these two great ones, which is more active in development, which has less bugs, which works best for us. At the end I choose `Leaflet.Editable`, but I really like the `Leaflet.Draw` functionality of drawing `Rectangle` in any direction and checking for self intersecting shapes (although I managed to achieve second one with `Leaflet.Editable` using `JSTS`).

Demo:
`Leaflet.Draw` https://drive.google.com/open?id=0B8DaJcGLql99NHJ5NDgtWk9WS2s
`Leaflet.Editable` https://drive.google.com/open?id=0B8DaJcGLql99YVB2bnJXVzZVdk0

**Comment by theashyster** [comment]
@yohanboniface Any thoughts?

**Comment by yohanboniface** [comment]
I'll have a look, but not in my top priorities atm sorry.

**Comment by theashyster** [comment]
Okay, thank you. I see, no problem.
I will probably try to invest a bit more time into this myself and try to find a solution, so if you have any pointers of direction that could help feel free to share. :)

**Comment by xoqem** [comment]
I commented in the duplicate I opened, but I have a fix for this, will put up a PR this afternoon or tomorrow.

**Comment by xoqem** [comment]
Here is my potential solution: https://github.com/Leaflet/Leaflet.Editable/pull/91

**Comment by theashyster** [comment]
@xoqem I used your code and tested this out a bit, worked fine with the initial drawing of the Rectangle. The drawing of Rectangle by click-click seemed a bit strange, but that is probably because before just dragging was possible, not this kind of way.

The thing that did not work though, when the drawing is already committed, then you once again get the same thing that once you try to drag one of the vertexes right across the others it moves the shape itself.

**Comment by xoqem** [comment]
@theashyster that is good feedback, thanks for sharing.  So in the app I'm using it for, I grab the shape after commit, add it to my own layer and clear the drawing layer, so that's why it works out in my case.

I'll try to spend some more cycles in the next week or so to find a fix, and will update if so.

Out of curiosity, does your app use the circle drawing?  I also tweaked it a bit in my PR and was curious how that worked out for other people's use cases.

**Comment by theashyster** [comment]
@xoqem :) I see, so you don't edit the drawing layer after you have committed it.

Okay, thanks for looking into this, I would have loved to look at it myself, but got my hands full in other tasks.

Nah, sadly it does not use Circle drawing, I saw the changes, but cannot test them in the system I am working with.

**Comment by xoqem** [comment]
@theashyster I looked into the scenario you mentioned, and it wasn't as bad as I feared.  Basically when editing an existing rectangle it behaves as it did before.  This is because when drawing the rectangle we have a vertex we can anchor (the first one they created).  When editing, its a little less clear if one of the vertices should be anchored (maybe the one in the opposite corner of the one you are dragging?).  Though, it seems in most cases when resizing an existing rectangle, you would be tweaking one of the sides, and less interested in crossing over, vs when you are first drawing the rectangle, you may want to go in any direction from the starting point.  If I have any brainstorms for changing the editing behavior I'll update, but it seems at least operable in its current state.

Also, you mentioned the 2 clicks when creating a rectangle, you can also click and drag as before, both should work.  I'm on the fence whether the 2 clicks behavior is preferable, though before a click and no drag would create a rectangle with no area, which seems unlikely to be the goal of the user, and allowing 2 clicks to create is more in line with how polygon works.  Though, I'm glad you mentioned, because that spurred me to try it on my iPhone and the 2 clicks approach didn't really work, though touch and drag still did.

**Comment by xoqem** [comment]
@theashyster Ok, I removed the method of adding a rectangle or circle with two clicks.  Like you mentioned, it was a little odd for the more normal use case you have (where you add shapes, then edit them).  It also fixes the touch screen problem I was having.

Still pondering if there is a nice solution when editing an existing rectangle to anchor one of the points.  Thanks so much for all the feedback on my PR, it helped me track down a couple of issues that would have popped up in my app eventually!

**Comment by theashyster** [comment]
@xoqem Great, thanks for looking into it. Yep, I know that and I am torn, I don't know if it should act as it was before or it should allow crossing over. If we take `Leaflet.draw` as an example again, then they allow it to cross over also on editing, that is why I think it should be that way. I think, that when editing the opposite vertex of the one that is dragged should be anchored, yes. Of course, yes, in most cases you would be tweaking sides and not care about crossing over, but still it would feel strange that when you start you can draw the rectangle in one way, then on editing it works differently.

I know, that I can still click and drag as before, just I was not used to the new behavior and it seemed strange (if it would be implemented like this from the start it would feel totally normal). But I agree with your points made, a `Rectangle` without an area also makes no sense, that is why in my app I override this and don't allow user to commit unless something is drawn. What I do is - on the first click I create the `Rectangle` on the map, but do not commit it yet, then user takes the vertex and drags it on `dragEnd` I commit the shape (I also have some `minSize` implementation, that is also why I do it this way). Yeah, I think the 2 clicks approach did not work for you on iPhone because you used `onMouseMove` to update the `Rectangle` which does not fire on touch).

I think that if they just click and don't drag we should not commit the shape and commit it only when user has dragged the vertex a bit.

Okay, good. I think anchoring the opposite corner should be fine and would work well. No problem, I am eager to help, because I would love for this to work in any direction. Even if it won't work as I would want at the end, I could take your code and try to tweak it myself later on, so thank you for this. :)

**Comment by xoqem** [comment]
Sure thing, glad I can help.

So I spent some more time poking into allowing the edit of an existing rectangle to drag past its opposite corner, and here is what I found:

The method that gets called for resizing an exiting rectangle is the extendBounds method in the RectangleEditor.  If you comment out this method, just out of curiosity, the dragging still works, but it behaves as if you are dragging the vertex of a polygon.  This does allow you to drag past any of the corners, but it doesn't keep the rectangle lines straight.  So, basically, the extendBounds method in RectangleEditor runs after the polygon style vertex drag, and fixes the edges to be straight like a rectangle, based on the bounds created by the vertex being dragged and the vertex in the opposite corner.

To update the shape when the extendsBounds logic is run, it calls updateBounds in the RectangleEditor.  The updatesBounds method takes the bounds, gets lat lng objects for the bounds using _boundsToLatLngs (so 1 lat lng object for each corner, for a total of 4) and updates the existing lat lng values of the shape.  This is where the problem eventually happens when you cross an edge.

Basically, the order of lat lng objects retured from _boundsToLatLngs consistently starts in one of the corners, and returns the lat lngs as it goes around the shape.  So, when you invert the shape the by dragging a vertex across the shape, the lat lng for the vertex that you are dragging is now returned in a different order from _boundsToLatLngs.  However, the vertex you are dragging is not in that order in the shape currently, so it starts updating other vertices to be at the position of the vertex you are dragging.  As the shape moves, it continually crosses either the horizontal or vertical edge, changing the order of vertices each time, which is why if you drag into the corner, they all move, as each pass a different vertex gets updated with the position of the dragged vertex.

I'm toying with a couple of ideas.

One options is to change updateBounds to preserve the order of vertices, though I need to play with this a bit and see if there is a reasonable way to do it that doesn't break other parts of rectangle drawing.

Another idea would be when the user crosses an edge to fire mouse up on the vertex they were dragging, and change the mouse event to start dragging the vertex they were crossing.  This doesn't feel like the ideal solution, as its basically hacking a lot of things with the mouse events.

The reason this doesn't happen when creating a rectangle is that the user isn't actually dragging a vertex in that scenario.  It stores where the location was when the mouse went down, and as the mouse moves it just sets the bounds to be between the location when the mouse went down, and the location the mouse is now.  Since its not updating a specific vertex, the order the lat lng values are in doesn't matter, so while the index of the vertex under the mouse will change, the code isn't actually doing anything with that vertex specifically.

**Comment by xoqem** [comment]
@theashyster Ok, I have a fix for editing an existing rectangle, turned out to not be too bad.  Basically, for extendBounds in rectangle, I don't call updateBounds, which can change the order of vertices.  Instead I manually set the 3 vertices, and don't touch the opposite vertex of the corner being dragged.  This seems to work fine, and I see no bad side effects from playing with so far.  Let me know what you think, and if you see any problems.  Thanks!

**Comment by theashyster** [comment]
@xoqem Thanks a lot for all of this, I have tested this a bit and it looks like it works as expected now, awesome job. :)

I see, heh, thanks for the explanations on how things work under the hood for `boundsToLatLngs`, I did not have the time to dig deeper into this myself, so reading an explanation so deep is really useful. A lot of this was known by me already, but I did not knew that without the `extendBounds` method it just draws like a `Polygon` and that this functions straightens the lines.

Okay, `boundsToLatLngs` is the reason, hmm, then I will have to find a new way of how and when to convert `LatLng` to different `CRS` that I am doing right now in this function, because we will not `updateBounds` in `extendBounds`. I will have to think about this a bit more. I actually had a logic where I pass in the vertex index that is dragged to `boundsToLatLngs` and then created bounds inside that function depending on which corner is dragged, so maybe something like that could be done, but I can test this out myself.

Something like this
`var pxBounds = vertexIndex === 1 || vertexIndex === 3 ? L.bounds(northWest, southEast) : L.bounds(southWest, northEast);`

Maybe I will be able to keep the order somehow and run `boundsToLatLngs`. I can probably just update each vertex in the feature in `extendBounds` as you do that, but also do the `CRS` projection and that would work. Or also give vertex order or vertex index to the `updateBounds` function to know the order it should keep.

Yeah, the drawing of `Rectangle` differs from other shapes, that it just extends the bounds all the time and does not draw anything before mouse is up again / the drag ends.

Great, yeah, I also did not see any bad side effects or problems with this, just that I have to adjust how I project my points on the map in different `CRS`, but I can tweak that, no problem. Thanks again, this is perfect.
